src/main/java/org/opensearch/sdk/ssl/SSLNettyTransport.java

```diff
@@ -0,0 +1,328 @@
+ */
+public class SSLNettyTransport extends Netty4Transport {
+
+    private static final Logger logger = LogManager.getLogger(SSLNettyTransport.class);
+    private final SslKeyStore ossks;
+
+    /**
+     *
+     * @param settings SSL Settings
+     * @param version version
+     * @param threadPool threadPool
+     * @param networkService networkService
+     * @param pageCacheRecycler pageCacheRecycler
+     * @param namedWriteableRegistry namedWriteableRegistry
+     * @param circuitBreakerService circuitBreakerService
+     * @param ossks ossks
+     * @param sharedGroupFactory sharedGroupFactory
+     */
+    public SSLNettyTransport(
+        final Settings settings,
+        final Version version,
+        final ThreadPool threadPool,
+        final NetworkService networkService,
+        final PageCacheRecycler pageCacheRecycler,
+        final NamedWriteableRegistry namedWriteableRegistry,
+        final CircuitBreakerService circuitBreakerService,
+        final SslKeyStore ossks,
+        SharedGroupFactory sharedGroupFactory
+    ) {
+        super(
+            settings,
+            version,
+            threadPool,
+            networkService,
+            pageCacheRecycler,
+            namedWriteableRegistry,
+            circuitBreakerService,
+            sharedGroupFactory
+        );
+
+        this.ossks = ossks;
+    }
+
+    /**
+     *
+     * @param channel The channel
+     * @param e Exception
+     */
+    @Override
+    public void onException(TcpChannel channel, Exception e) {
+
+        Throwable cause = e;
+
+        if (e instanceof DecoderException && e != null) {
+            cause = e.getCause();
+        }
+
+        logger.error("Exception during establishing a SSL connection: " + cause, cause);
+
+        super.onException(channel, e);
+    }
+
+    /**
+     *
+     * @param name name of channel
+     * @return ChannelHandler
+     */
+    @Override
+    protected ChannelHandler getServerChannelInitializer(String name) {
+        return new SSLServerChannelInitializer(name);
+    }
+
+    /**
+     *
+     * @param node Node this channel is connected to
+     * @return ChannelHandler
+     */
+    @Override
+    protected ChannelHandler getClientChannelInitializer(DiscoveryNode node) {
+        return new SSLClientChannelInitializer(node);
+    }
+
+    /**
+     * SSLServerChannelInitializer
+     */
+    protected class SSLServerChannelInitializer extends Netty4Transport.ServerChannelInitializer {
+
+        /**
+         *
+         * @param name name of this channel
+         */
+        public SSLServerChannelInitializer(String name) {
+            super(name);
+        }
+
+        /**
+         *
+         * @param ch            the {@link Channel} which was registered.
+         * @throws Exception
+         */
+        @Override
+        protected void initChannel(Channel ch) throws Exception {
+            super.initChannel(ch);
```

Comment:
I see security plugin has a dual mode here. Can you help me understand what's the significance of dual mode?

Reply:
Dual means that it supports encrypted and unencrypted traffic at the same time. To be honest, I have never used this feature of the security plugin before and it needs to be explicitly turned on. https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/ssl/transport/DualModeSSLHandler.java#L69-L75